### PR TITLE
AEROGEAR-7555 Refactor client CRULD operations

### PR DESCRIPTION
**Pull request by vchepeli**

### CODE REVIEW

File: pkg/keycloak/client.go [hunk 1]
```diff
@@ -38,637 +36,368 @@ type Client struct {
...
+		err := json.Unmarshal(body, &users)
+		return users, err
+	})
+	return result.([]*v1alpha1.KeycloakUser), err
+}
 
-	return nil
+func (c *Client) ListIdentityProviders(realmName string) ([]*v1alpha1.KeycloakIdentityProvider, error) {
+	result, err := c.list(fmt.Sprintf("realms/%s/identity-provider/instances", realmName), "identity providers", func(body []byte) (T, error) {
+		var users []*v1alpha1.KeycloakUser
```
Comment (dimitraz) [inline pkg/keycloak/client.go]: Can you change this to `var providers []*v1alpha1.KeycloakIdentityProvider`?

File: pkg/keycloak/client.go [hunk 2]
```diff
@@ -38,637 +32,368 @@ type Client struct {
...
+	return result.(*v1alpha1.KeycloakRealm), err
+}
 
-	req.Header.Add("Authorization", "Bearer "+c.token)
-	res, err := c.requester.Do(req)
-	if err != nil {
-		logrus.Errorf("error on request %+v", err)
-		return errors.Wrap(err, "error performing delete client request")
-	}
+func (c *Client) GetClient(clientId, realmName string) (*v1alpha1.KeycloakClient, error) {
```
Comment (dimitraz) [inline pkg/keycloak/client.go]: Can you change all `clientId`s and `userId`s to `clientID` etc for consistency?

File: pkg/keycloak/client.go [hunk 3]
```diff
@@ -38,637 +32,368 @@ type Client struct {
...
-	}
+func (c *Client) UpdateClient(kcClient, objClient *v1alpha1.KeycloakClient, realmName string) error {
+	err := c.update(objClient, fmt.Sprintf("realms/%s/clients/%s", realmName, kcClient.ID), "client")
+	return err
+}
 
-	if res.StatusCode < 201 || res.StatusCode > 409 {
-		return errors.New("failed to create user: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
-	}
+func (c *Client) UpdateUser(kcUser, objUser *v1alpha1.KeycloakUser, realmName string) error {
```
Comment (dimitraz) [inline pkg/keycloak/client.go]: Can you change all `objUser`s and `objClient`s to `specUser` etc for consistency?

File: pkg/keycloak/client.go [hunk 4]
```diff
@@ -38,637 +32,368 @@ type Client struct {
...
 
-	return nil
+func (c *Client) CreateIdentityProvider(identityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error {
+	err := c.create(identityProvider, fmt.Sprintf("realms/%s/identity-provider/instances", realmName), "identity provider")
+	return err
 }
 
-func (c *Client) GetClient(clientID string, realmName string) (*v1alpha1.KeycloakClient, error) {
+//================================================= READ =================================================
+func (c *Client) get(resourcePath, resourceName string, unMarshallFunc func(body []byte) (T, error)) (T, error) {
```
Comment (dimitraz) [inline pkg/keycloak/client.go]: Can you change these func names to 'unmarshalFunc`?

**Review by dimitraz — COMMENTED** [review]
Looks good so far just a few comments

File: pkg/keycloak/client.go (hunk 3 above)
Comment (mikenairn) [reply pkg/keycloak/client.go]: Why is kcUser here? Thought we removed that.

File: pkg/keycloak/client.go [hunk 5]
```diff
@@ -38,637 +32,368 @@ type Client struct {
...
+		err := json.Unmarshal(body, user)
+		return user, err
+	})
+	return result.(*v1alpha1.KeycloakUser), err
+}
 
-	return nil
+func (c *Client) GetIdentityProvider(alias string, realmName string) (*v1alpha1.KeycloakIdentityProvider, error) {
+	result, err := c.get(fmt.Sprintf("realms/%s/identity-provider/instances/%s", realmName, alias), "identity provider", func(body []byte) (T, error) {
+		var user *v1alpha1.KeycloakIdentityProvider
```
Comment (dimitraz) [inline pkg/keycloak/client.go:166]: maybe `provider`/`providers` are better for these var names

File: pkg/keycloak/client.go [hunk 6]
```diff
@@ -678,25 +403,27 @@ func defaultRequester() Requester {
 }
 
 type KeycloakInterface interface {
-	ListRealms() ([]*v1alpha1.KeycloakRealm, error)
-	GetRealm(realmName string) (*v1alpha1.KeycloakRealm, error)
 	CreateRealm(realm *v1alpha1.KeycloakRealm) error
-	UpdateRealm(realm *v1alpha1.KeycloakRealm) error
+	GetRealm(realmName string) (*v1alpha1.KeycloakRealm, error)
+	UpdateRealm(kcRealm, objRealm *v1alpha1.KeycloakRealm) error
```
Comment (mikenairn) [inline pkg/keycloak/client.go]: Why do we need both kcRealm and objRealm here?

Comment (dimitraz) [inline pkg/keycloak/client.go]: I think the signature of these update functions are wrong. We removed the need for the `kc` object by adding any necessary fields to the `spec` object (like the ID) before calling the update function

**Review by dimitraz — CHANGES_REQUESTED** [review]

File: pkg/keycloak/client.go [hunk 7]
```diff
@@ -678,25 +403,27 @@ func defaultRequester() Requester {
...
 	DeleteUser(userID, realmName string) error
-	UpdateUser(specUser *v1alpha1.KeycloakUser, realmName string) error
 	ListUsers(realmName string) ([]*v1alpha1.KeycloakUser, error)
 
 	CreateIdentityProvider(identityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error
 	GetIdentityProvider(alias, realmName string) (*v1alpha1.KeycloakIdentityProvider, error)
-	UpdateIdentityProvider(identityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error
+	UpdateIdentityProvider(kcIdentityProvider, specIdentityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error
 	DeleteIdentityProvider(alias, realmName string) error
 	ListIdentityProviders(realmName string) ([]*v1alpha1.KeycloakIdentityProvider, error)
```
Comment (mikenairn) [inline pkg/keycloak/client.go]: I don't think the keycloak client needs to be aware of spec and kc objects, we should just pass one object in that is the thing we want to update.

If IDs or whatever really need to be passed in they should be passed explicitly, but we intended to have the spec object have all fields set from the kc object so it should have everything needed to create a request.

**Review by dimitraz — APPROVED** [review]
Verified! :+1:

**Pull request merged**